### ISSUE TITLE
Bounds

### ISSUE THREAD

**Issue by webdesserts** [webdesserts/alchemist-js#1]
How should alchemist handle boundries and clipping?

The end user api might look something like this:

```js
alchemize.rgb(255, 255, 255) // => Color
alchemize.rgb(256, 255, 255) // => null
```

what the plugin interface might look like:

```js
alchelmize.use({
  name: 'rgb',
  bounds: function bounds (r, g, b) {
    // success?
    return true
    // clipped?
    return [r, g, b]
    // invalid?
    return false
  },
  to: function () {}
})
```
Ya, that's pretty messy. If there's a way that I could make that cleaner, it would be nice. I'm not even sure if I want to allow clipping yet. The problem though, is that if I don't implement it inside the library, the color space developer would implement it inside their conversion functions and the user would have no way to turn the clipping off if they just want a null value when something fails.

This also brings up the issue of wether or not we should return a soft error as null, or actually throw an error and let the user handle it with something like promises. Maybe we should make this a configuration option?

```js
  Alchemist.create({bounds: true}).rgb(256,255,255) // => null (the default)
  Alchemist.create({bounds: 'strict'}).rgb(256,255,255) // Throws Error
  Alchemist.create({bounds: 'clip'}).rgb(256,256,256) // => [255,255,255]
```
...aaaand finally this brings up the issue of whether we should clip on color interpretation or just color conversion?

```js
  var alchemize() = Alchemist.create({bounds: 'clip'})
  alchemize.rgb(256, 255, 255) // Seems like an error would be expected
  alchemize.lch(200, 200, 200).rgb() // => [91, 50, 197] (clipped value)
```

**Comment by webdesserts** [comment]
Closed by c3cf3bd6d2b57593b12c22886e02e5938c0b7a54 :tada:

**Comment by webdesserts** [comment]
For anyone wondering, the new api looks like this:

### Settings
```javascript
alchemist.init({ limits: 'clip' })
```

### Plugins
```javascript
var rgb = {
  name: 'rgb',
  limits: {
    max: [255, 255, 255],
    min: [0, 0, 0]
  }
  to: {}
}
```
